File: Linked_List/Reverse_Linked_List/Reverse_Linked_List.java
// Leetcode: https://leetcode.com/problems/reverse-linked-list/

/*
        // Definition for Singly-Linked List
        public class ListNode
	{
                int val;
                ListNode next;
                public ListNode()
                {
                }
                public ListNode( int val )
                {
                        this.val = val;
                }
                public ListNode( int val, ListNode next )
                {
                        this.val = val;
                        this.next = next;
                }
        }
*/

class Solution
{
        public ListNode reverseList( ListNode head )
        {
		if( head == null )
		{
			return null;
		}

		ListNode previous = null;
		ListNode current = head;

		while( current != null )
		{
			ListNode next = current.next;
			current.next = previous;
			previous = current;
			current = next;
		}

		return head = previous;
        }
}

File: Linked_List/Reverse_Linked_List/Reverse_Linked_List.js
// Leetcode: https://leetcode.com/problems/reverse-linked-list/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */

const reverseList = (head) => {
        if (head == null) {
                return null;
        }

        let previous = null;
        let current = head;

        while (current != null) {
                const next = current.next;
                current.next = previous;
                previous = current;
                current = next;
        }

        return head = previous;
};